### PR TITLE
Add Settings API for servers, zones, and regions

### PR DESCRIPTION
**Pull request by chrisarcand**
This adds manipulation of all settings on servers, zones, and regions via the API in core's `ConfigurationManagementMixin` in the form of a new base controller method, `settings`, which routes to these three collections using the `settings` options keyword in the YAML.

* `GET /servers/:id/settings`
  Renders the server settings as provided by the settings API.

* `PATCH /zones/:id/settings`, body: `{"ems": { "ems_ansible_tower": "blacklisted_event_names": ["EventName1"] } }`
  Sets the ems.ems_ansible_tower.blacklisted_event_names to ["EventName1"] This action serves as both create and update.

* `DELETE /zones/:id/settings`, body: `["ems", "ems_ansible_tower"]`
  This deletes the settings change for this zone under this key path. All changes underneath this key path would be reset and the settings for this zone's ems.ems_ansible_tower subtree would revert to whatever is higher up the settings tree for this zone.

The interface here is very simplistic and just follows the API provided by the core mixin. It emulates the existing `/settings` endpoint.

Additionally,

* Only super admins can read, edit, and delete all settings
* All other users are read only and settings are scoped to only those set as categories in the settings collection.

https://bugzilla.redhat.com/show_bug.cgi?id=1486634

### CODE REVIEW

**Comment by chrisarcand** [comment]
cc/ @gtanzillo

**Comment by chrisarcand** [comment]
Ready for review; the remaining rubocop stuff is exempt in RSpec DSL, IMO.

**Comment by abellotti** [comment]
Thanks @chrisarcand for the quick turnaround with this PR 👍 

just a couple of notes:

- You probably meant DELETE instead of DESTROY in the description 😉 
- Non super-admin users should minimally have the ops_setting role to be able to read settings
- GETs of /api/settings should mimic GET /api/servers/:myserver_id/settings, so for super-admin, return all settings there and not just the white-listed ones.

Other things:

- would have been nice to also support
  - POST/PATCH/DELETE /api/settings/path/
  - POST/PATCH/DELETE /api/[zones|regions|servers]/:id/settings/path/

  but that can be a follow-up PR.

**Comment by chrisarcand** [comment]
> You probably meant DELETE instead of DESTROY in the description

Ah ha, yup. 😄 

> Non super-admin users should minimally have the ops_setting role to be able to read settings

Ok! Will add

> GETs of /api/settings should mimic GET /api/servers/:myserver_id/settings, so for super-admin, return all settings there and not just the white-listed ones.

Ah yeah, I noticed that as well. However, the API that was built for that endpoint for some reason doesn't use the settings API and instead digs directly in to the database and parses records itself (and ignores the yaml layer), so I wasn't going to open that can of worms here. I guess I can correct it here though, that's cool 👍 

> would have been nice to also support
>    POST/PATCH/DELETE /api/settings/path/
>    POST/PATCH/DELETE /api/[zones|regions|servers]/:id/settings/path/

* This route is not RESTful, intentionally. The difference between POST and PATCH doesn't mean anything because from the view of the consumer, you're always *editing* a resource's settings (that the actual action you're taking is entering a record in the database for a particular server to override the environment settings is an implementation detail). In the case of adding a new settings that wasn't an override (like anything. { "my_new_setting" => true }), it's not worth being pedantic to the consumer by saying 'When it's an existing entry in the hash you should use PATCH but when it's a completely new one that isn't defaulted you should use POST'. So no POSTs, is my point here.

* I originally had /path stuff so you could do arbitrary subtrees like `/api/settings/ems/ems_ansible_tower/` but decided that it doesn't actually add that much value for the consumer and then handcuffs us as *no* new paths could be added to this endpoint since a user can arbitrarily add any settings. I guess I'm not completely against it but I just didn't see the value. 

Perhaps we should actually ask consumers of the API if they think that would be helpful, as someone like QE for example might just want all the settings back anyway. 


Addressing actionable feedback now! ❤️

**Comment by chrisarcand** [comment]
@abellotti Ready for 👀 again

**Comment by abellotti** [comment]
Thanks @chrisarcand for the updates. Lookin' good 👍 

As chatted in Gitter, we'd need a follow-up in Settings API land to not allow disabling web_services in servers role if this was the last server.

**Comment by chrisarcand** [comment]
:shipit:

File: spec/requests/regions_spec.rb [hunk 1]
```diff
@@ -40,4 +40,112 @@
...
+    # let(:server) { FactoryGirl.create(:miq_server) }
+    # let(:region) { FactoryGirl.create(:miq_region, :region => FactoryGirl.create(:miq_region, :region => ApplicationRecord.my_region_number)) }
+    let(:region) { @region }
+
+    let(:original_timeout) { region.settings_for_resource[:api][:authentication_timeout] }
+    let(:super_admin) { FactoryGirl.create(:user, :role => 'super_administrator', :userid => 'alice', :password => 'alicepassword') }
+
+    # before do
+    #   allow(MiqServer).to receive(:in_region).with(ApplicationRecord.my_region_number).and_return(MiqServer.where(:id => server.id))
+    # end
```
Comment (abellotti) [inline spec/requests/regions_spec.rb]: @imtayadeway could you remove the commented code above ( & lines 45-46) and repush ?  unless you needed this for reference 😉

**Comment by chrisarcand** [comment]
@abellotti I think your comment above is on old changes; hit refresh 😄

**Comment by abellotti** [comment]
LGTM!! Thanks @chrisarcand and @imtayadeway for this greatly needed enhancement !!! 😍

**Comment by Fryguy** [comment]
Really nice work @chrisarcand !!

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit acb952b9e4bcedde51b085bcf2edb1d4f588376b
Author: Alberto Bellotti <abellotti@users.noreply.github.com>
Date:   Mon Jan 15 17:08:07 2018 -0500

    Merge pull request #275 from chrisarcand/settings-crud
    
    Add Settings API for servers, zones, and regions
    (cherry picked from commit 3f407638b474b1ff8b2945e9a13666d1fb061201)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1535602
```

**Comment by simaishi** [comment]
Fine backport (to manageiq repo) details:

```
$ git log -1
commit e4138807e933225b3e414aab90eb8b2594c082b9
Author: Alberto Bellotti <abellotti@users.noreply.github.com>
Date:   Mon Jan 15 17:08:07 2018 -0500

    Merge pull request #275 from chrisarcand/settings-crud
    
    Add Settings API for servers, zones, and regions
    (cherry picked from commit 3f407638b474b1ff8b2945e9a13666d1fb061201)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1535605
```

**Pull request merged**